Kitodo/src/main/java/org/kitodo/services/image/ImageGenerator.java

```diff
@@ -0,0 +1,423 @@
+import java.util.stream.Stream;
+
+import javax.imageio.ImageIO;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kitodo.api.filemanagement.FileManagementInterface;
+import org.kitodo.api.imagemanagement.ImageFileFormat;
+import org.kitodo.api.imagemanagement.ImageManagementInterface;
+import org.kitodo.config.xml.fileformats.FileFormat;
+import org.kitodo.data.database.beans.Folder;
+import org.kitodo.helper.tasks.EmptyTask;
+import org.kitodo.model.Subfolder;
+import org.kitodo.production.thread.TaskImageGeneratorThread;
+import org.kitodo.production.thread.TaskScriptThread;
+import org.kitodo.serviceloader.KitodoServiceLoader;
+
+/**
+ * A program that generates images using the image management interface. This
+ * program is run by the {@link TaskImageGeneratorThread} when the user manually
+ * initiates the creation of the images. If the images are generated when the
+ * task is completed, this is done by the {@link TaskScriptThread}.
+ */
+public class ImageGenerator implements Runnable {
+    private static final Logger logger = LogManager.getLogger(ImageGenerator.class);
+
+    /**
+     * Output folders.
+     */
+    private Collection<Subfolder> outputs;
+
+    /**
+     * Current position in list.
+     */
+    private int position;
+
+    /**
+     * Folder with source images.
+     */
+    private Subfolder sourceFolder;
+
+    /**
+     * List of possible source images.
+     */
+    private List<Pair<String, URI>> sources;
+
+    /**
+     * Current step of the generation process.
+     */
+    private ImageGeneratorStep state;
+
+    /**
+     * Task in the TaskManager that runs this ImageGenerator.
+     */
+    private Optional<EmptyTask> supervisor = Optional.empty();
+
+    /**
+     * List of elements to be generated.
+     */
+    private List<Pair<Pair<String, URI>, List<Subfolder>>> toBeGenerated;
+
+    /**
+     * Variant of image generation, see there.
+     */
+    private GenerationMode mode;
+
+    /**
+     * Creates a new image generator.
+     *
+     * @param sourceFolder
+     *            image source folder
+     * @param mode
+     *            whether all, or only a few images are to be generated
+     * @param outputs
+     *            output folders to generate to
+     */
+    public ImageGenerator(Subfolder sourceFolder, GenerationMode mode, Collection<Subfolder> outputs) {
+        this.sourceFolder = sourceFolder;
+        this.mode = mode;
+        this.outputs = outputs;
+        this.state = ImageGeneratorStep.LIST_SOURCE_FOLDER;
+        this.sources = Collections.emptyList();
+        this.toBeGenerated = new LinkedList<>();
+    }
+
+    /**
+     * Appends the element to the list of elements to be generated.
+     *
+     * @param elementToAppend
+     *            element to be appended to the list
+     */
+    void addToToBeGenerated(Pair<Pair<String, URI>, List<Subfolder>> elementToAppend) {
```

Comment:
Please create a class, instead of using Pair<Pair<String, URI>, List<Subfolder>>